src/NumericalAlgorithms/RootFinding/GslMultiRoot.hpp

```diff
@@ -0,0 +1,364 @@
+    double relative_tolerance, Verbosity verbosity,
+    double maximum_absolute_tolerance, SolverType solver_type,
+    StoppingCondition condition, SolverAlloc solver_alloc, SolverSet solver_set,
+    SolverIterate solver_iterate, SolverFree solver_free) {
+  // Supply gsl_root with the initial guess:
+  gsl_vector* const gsl_root = gsl_vector_alloc(Dim);
+  gsl_vector_set_with_std_array(gsl_root, initial_guess);
+  auto* solver = solver_alloc(solver_type, Dim);
+  solver_set(solver, &f, gsl_root);
+
+  // Take iterations:
+  if (verbosity == Verbosity::Verbose or verbosity == Verbosity::Debug) {
+    Parallel::printf("\nAttempting a root find...\n");
+  }
+  int status;
+  size_t iteration_number = 0;
+  do {
+    if (verbosity == Verbosity::Debug) {
+      print_state<Dim>(iteration_number, solver, iteration_number == 0);
+    }
+    iteration_number++;
+    status = solver_iterate(solver);
+    // Check if solver is stuck
+    if (status == GSL_ENOPROG) {
+      break;
+    }
+    if (condition == StoppingCondition::TestDelta) {
+      status = gsl_multiroot_test_delta(solver->dx, solver->x,
+                                        absolute_tolerance, relative_tolerance);
+    } else if (condition == StoppingCondition::TestResidual) {
+      // NOTE: Sometimes hybridsj works only with the test_residual condition
+      status = gsl_multiroot_test_residual(solver->f, absolute_tolerance);
+    } else {
+      ERROR(
+          "Invalid stopping condition. Has to be either TestDelta"
+          "or TestResidual.");
+    }
+  } while (status == GSL_CONTINUE and iteration_number < maximum_iterations);
+  if (verbosity == Verbosity::Verbose or verbosity == Verbosity::Debug) {
+    print_state<Dim>(iteration_number, solver, verbosity == Verbosity::Verbose);
+  }
+  bool success = (status == GSL_SUCCESS);
+  if (verbosity != Verbosity::Silent) {
+    Parallel::printf("\n");
+    if (not success) {
+      const std::string ascii_divider = std::string(80, '#');
+      const std::string failure_message =
+          ascii_divider + "\n\t\t\tWARNING: Root Finding FAILED\n" +
+          ascii_divider;
+      Parallel::printf("%s\n", failure_message);
+    } else {
+      Parallel::printf("Root finder converged.\n");
+    }
+  }
+  // If maximum_absolute_tolerance is given, return success = true
+  // as long as maximum_absolute_tolerance is achieved even if the
+  // root finder doesn't converge.
+  bool success_with_tolerance = true;
+  if (not success and maximum_absolute_tolerance > 0) {
+    for (size_t i = 0; i < Dim; ++i) {
+      if (fabs(gsl_vector_get(solver->f, i)) > maximum_absolute_tolerance) {
+        success_with_tolerance = false;
+      }
+    }
+  }
+  if (success_with_tolerance) {
+    success = true;
+  }
+
+  // Store the converged root in result
+  std::array<double, Dim> result = gsl_to_std_array<Dim>(solver->x);
+  solver_free(solver);
+  gsl_vector_free(gsl_root);
+
+  if (not success) {
+    throw rootfinding_exception{};
```

Comment:
Here please provide info as to *why* the root find failed. It might even be better to print all the debug info into a single std::string and then print that at the end. This way if there are multiple root finds ever going on at once their info won't get jumbled.